dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileSchema.cpp

```diff
@@ -39,58 +42,90 @@ String ColumnFileSchema::toString() const
+
+        const Holder & holder = iter->second;
+        if (auto p = holder.column_file_schema.lock(); p)
         {
-            if (iter->second.expired())
-            {
-                iter = column_file_schemas.erase(iter);
-            }
-            else
-            {
-                ++iter;
-            }
+            // If when the item is evicted, but the item is still be used by some ColumnFiles,
+            // we increment this metrics, to show maybe the max_size of SharedBlockSchemas is not enough now.
+            GET_METRIC(tiflash_shared_block_schemas, type_still_used_when_evict).Increment();
         }
-        return true;
-    },
-                                     /*multi*/ false,
-                                     /*interval_ms*/ 60000);
-}
 
-SharedBlockSchemas::~SharedBlockSchemas()
-{
-    if (handle)
+        --current_size;
+        lru_queue.pop_front();
+        column_file_schemas.erase(iter);
+        GET_METRIC(tiflash_shared_block_schemas, type_current_size).Decrement();
+    }
+
+    if (current_size > (1ull << 63))
     {
-        background_pool.removeTask(handle);
+        throw Exception(String(__FUNCTION__) + " inconsistent, current_size < 0", ErrorCodes::LOGICAL_ERROR);
     }
 }
 
+
 ColumnFileSchemaPtr SharedBlockSchemas::find(const Digest & digest)
 {
-    std::lock_guard<std::mutex> lock(mutex);
-    auto it = column_file_schemas.find(digest);
-    if (it == column_file_schemas.end())
+    std::lock_guard lock(mutex);
+    auto iter = column_file_schemas.find(digest);
+    if (iter == column_file_schemas.end())
+    {
         return nullptr;
-    return it->second.lock();
+    }
+
+    lru_queue.splice(lru_queue.end(), lru_queue, iter->second.queue_it);
+
+    return iter->second.column_file_schema.lock();
 }
 
 ColumnFileSchemaPtr SharedBlockSchemas::getOrCreate(const Block & block)
 {
     Digest digest = hashSchema(block);
-    std::lock_guard<std::mutex> lock(mutex);
-    auto it = column_file_schemas.find(digest);
-    if (it == column_file_schemas.end() || it->second.expired())
+
+    std::lock_guard lock(mutex);
+    auto iter = column_file_schemas.find(digest);
+    if (iter != column_file_schemas.end())
     {
-        auto schema = std::make_shared<ColumnFileSchema>(block);
-        column_file_schemas.emplace(digest, schema);
-        return schema;
+        if (auto schema = iter->second.column_file_schema.lock(); schema)
+        {
+            GET_METRIC(tiflash_shared_block_schemas, type_hit_count).Increment();
+            lru_queue.splice(lru_queue.end(), lru_queue, iter->second.queue_it);
+            return schema;
+        }
+        else
+        {
+            // if the weak_ptr.lock() only get nullptr, that means the schema is not used by any ColumnFiles
+            // So we need update the item
+            GET_METRIC(tiflash_shared_block_schemas, type_miss_count).Increment();
+            auto new_schema = std::make_shared<ColumnFileSchema>(block);
```

Comment:
Seems we should use `std::shared_ptr<ColumnFileSchema> new_schema(new ColumnFileSchema(block));` instead of `make_shared` when we need it to be released ASAP https://stackoverflow.com/questions/32113594/weak-ptr-make-shared-and-memory-deallocation

Reply:
<details>
<summary> test codes </summary>
```

//////////////////////////////////////////////////////////////////////////////
//
// process_mem_usage(double &, double &) - takes two doubles by reference,
// attempts to read the system-dependent data for a process' virtual memory
// size and resident set size, and return the results in KB.
//
// On failure, returns 0.0, 0.0
void process_mem_usage(double & vm_usage, double & resident_set)
{
    using std::ifstream;
    using std::ios_base;
    using std::string;

    vm_usage = 0.0;
    resident_set = 0.0;

    // 'file' stat seems to give the most reliable results
    //
    ifstream stat_stream("/proc/self/stat", ios_base::in);

    // dummy vars for leading entries in stat that we don't care about
    //
    string pid, comm, state, ppid, pgrp, session, tty_nr;
    string tpgid, flags, minflt, cminflt, majflt, cmajflt;
    string utime, stime, cutime, cstime, priority, nice;
    string o, itrealvalue, starttime;

    // the two fields we want
    //
    UInt64 vsize;
    Int64 rss;

    stat_stream >> pid >> comm >> state >> ppid >> pgrp >> session >> tty_nr
        >> tpgid >> flags >> minflt >> cminflt >> majflt >> cmajflt
        >> utime >> stime >> cutime >> cstime >> priority >> nice
        >> o >> itrealvalue >> starttime >> vsize >> rss; // don't care about the rest

    stat_stream.close();

    Int64 page_size_kb = sysconf(_SC_PAGE_SIZE) / 1024; // in case x86-64 is configured to use 2MB pages
    vm_usage = vsize / 1024.0;
    resident_set = rss * page_size_kb;
}

struct BigMem
{
    // 1mb
    char buff[1024 * 1024];
};

const size_t n_blocks = 10240;
static constexpr double step = 1024;

TEST(WeakPtr, MakeShared)
{
    std::vector<std::shared_ptr<BigMem>> shared_blocks;
    shared_blocks.reserve(n_blocks);
    std::vector<std::weak_ptr<BigMem>> weak_blocks;
    weak_blocks.reserve(n_blocks);

    double vm_kb, rss_kb;

    process_mem_usage(vm_kb, rss_kb);
    LOG_INFO(Logger::get(), "vm={:.3f}MB rss={:.3f}MB -- init", vm_kb / step, rss_kb / step);

    for (size_t i = 0; i < n_blocks; ++i)
    {
        auto p = std::make_shared<BigMem>();
        memset(p->buff, 0xff, 1024 * 1024);
        shared_blocks.emplace_back(p);
        weak_blocks.emplace_back(p);
    }
    process_mem_usage(vm_kb, rss_kb);
    LOG_INFO(Logger::get(), "vm={:.3f}MB rss={:.3f}MB -- allocated", vm_kb / step, rss_kb / step);

    for (size_t i = 0; i < n_blocks; ++i)
    {
        shared_blocks[i].reset();
    }
    process_mem_usage(vm_kb, rss_kb);
    LOG_INFO(Logger::get(), "vm={:.3f}MB rss={:.3f}MB -- shared_ptrs realeased", vm_kb / step, rss_kb / step);

    for (size_t i = 0; i < n_blocks; ++i)
    {
        weak_blocks[i].reset();
    }
    process_mem_usage(vm_kb, rss_kb);
    LOG_INFO(Logger::get(), "vm={:.3f}MB rss={:.3f}MB -- weak_ptrs released", vm_kb / step, rss_kb / step);
}

TEST(WeakPtr, NewThenShared)
{
    std::vector<std::shared_ptr<BigMem>> shared_blocks;
    shared_blocks.reserve(n_blocks);
    std::vector<std::weak_ptr<BigMem>> weak_blocks;
    weak_blocks.reserve(n_blocks);

    double vm_kb, rss_kb;

    process_mem_usage(vm_kb, rss_kb);
    LOG_INFO(Logger::get(), "vm={:.3f}MB rss={:.3f}MB -- init", vm_kb / step, rss_kb / step);

    for (size_t i = 0; i < n_blocks; ++i)
    {
        std::shared_ptr<BigMem> p(new BigMem());
        memset(p->buff, 0xff, 1024 * 1024);
        shared_blocks.emplace_back(p);
        weak_blocks.emplace_back(p);
    }
    process_mem_usage(vm_kb, rss_kb);
    LOG_INFO(Logger::get(), "vm={:.3f}MB rss={:.3f}MB -- allocated", vm_kb / step, rss_kb / step);

    for (size_t i = 0; i < n_blocks; ++i)
    {
        shared_blocks[i].reset();
    }
    process_mem_usage(vm_kb, rss_kb);
    LOG_INFO(Logger::get(), "vm={:.3f}MB rss={:.3f}MB -- shared_ptrs realeased", vm_kb / step, rss_kb / step);

    for (size_t i = 0; i < n_blocks; ++i)
    {
        weak_blocks[i].reset();
    }
    process_mem_usage(vm_kb, rss_kb);
    LOG_INFO(Logger::get(), "vm={:.3f}MB rss={:.3f}MB -- weak_ptrs released", vm_kb / step, rss_kb / step);
}
```
</details>

```
[==========] Running 2 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 2 tests from WeakPtr
[ RUN      ] WeakPtr.MakeShared
[2023/02/25 17:32:36.089 +08:00] [INFO] [gtest_s3gcmanager.cpp:322] ["vm=863.047MB rss=44.629MB -- init"] [thread_id=1]
[2023/02/25 17:32:45.105 +08:00] [INFO] [gtest_s3gcmanager.cpp:332] ["vm=14179.047MB rss=12910.754MB -- allocated"] [thread_id=1]
[2023/02/25 17:32:45.126 +08:00] [INFO] [gtest_s3gcmanager.cpp:339] ["vm=14179.047MB rss=12910.789MB -- shared_ptrs realeased"] [thread_id=1]
[2023/02/25 17:32:55.808 +08:00] [INFO] [gtest_s3gcmanager.cpp:346] ["vm=14179.047MB rss=82.699MB -- weak_ptrs released"] [thread_id=1]
[       OK ] WeakPtr.MakeShared (19719 ms)
[ RUN      ] WeakPtr.NewThenShared
[2023/02/25 17:32:55.809 +08:00] [INFO] [gtest_s3gcmanager.cpp:359] ["vm=14179.047MB rss=82.699MB -- init"] [thread_id=1]
[2023/02/25 17:33:03.688 +08:00] [INFO] [gtest_s3gcmanager.cpp:369] ["vm=14179.047MB rss=10350.934MB -- allocated"] [thread_id=1]
[2023/02/25 17:33:06.908 +08:00] [INFO] [gtest_s3gcmanager.cpp:376] ["vm=14179.047MB rss=254.434MB -- shared_ptrs realeased"] [thread_id=1]
[2023/02/25 17:33:06.933 +08:00] [INFO] [gtest_s3gcmanager.cpp:383] ["vm=14179.047MB rss=247.406MB -- weak_ptrs released"] [thread_id=1]
[       OK ] WeakPtr.NewThenShared (11126 ms)
[----------] 2 tests from WeakPtr (30845 ms total)
```

Reply:
https://lanzkron.wordpress.com/2012/04/22/make_shared-almost-a-silver-bullet/